calyx/src/ir/component.rs

```diff
@@ -103,3 +98,84 @@ impl Component {
         self.namegen.gen_name(prefix)
     }
 }
+
+/// A wrapper struct exposing an ordered collection of named entities within an
+/// RRC with deterministic iteration and constant-time look-up on names
+/// directly. The struct assumes that the name of an entity cannot change. Doing
+/// so will introduce incorrect results for look-ups.
+#[derive(Debug)]
+pub struct IdList<T: GetName>(LinkedHashMap<Id, RRC<T>>);
+
+impl<T: GetName> IdList<T> {
+    /// Removes all elements from the collection
+    pub fn clear(&mut self) {
+        self.0.clear();
+    }
+
+    /// Keep only the elements in the collection which satisfy the given
+    /// predicate
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&RRC<T>) -> bool,
+    {
+        for entry in self.0.entries() {
+            if !f(entry.get()) {
+                entry.remove();
+            }
+        }
+    }
+
+    /// Add a new element to the colleciton
+    pub fn add(&mut self, item: RRC<T>) {
+        let name = item.clone_name();
+        self.0.insert(name, item);
+    }
+
+    /// Add multiple elements to the collection from an owned vector
+    pub fn add_multiple(&mut self, items: Vec<RRC<T>>) {
```

Comment:
Rust's iter canonically has `extend` to extend a collection with the contents of another collection: https://doc.rust-lang.org/std/iter/trait.Extend.html

Maybe we can use the same name unless there is a good reason not to?

Reply:
sure, I can implement this as extend instead, that's nice